lib/mongo/server/connection.rb

```diff
@@ -190,9 +190,9 @@ def handshake!
 
       def authenticate!
         if options[:user] || options[:auth_mech]
-          user = Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
+          @user ||= Auth::User.new(Options::Redacted.new(:auth_mech => default_mechanism).merge(options))
```

Comment:
Thanks for the info; I've updated it to properly handle this.

Additionally, there was an issue with the salted password not being able to be calculated when the client key was cached because it was no longer needed until generating the server key, and by that time a new reply had come since the one that contained the salt. I fixed this by explicitly calling `salted_password` in `SCRAM::Conversation#continue`.